File: android/app/src/main/java/com/nexo/app/domain/model/WorkoutLog.kt
package com.nexo.app.domain.model

/**
 * Mirrors `gyms/{gymId}/members/{uid}/workoutLogs/{logId}` — see
 * FIRESTORE_SCHEMA.md. `id` is a client-generated UUID string, matching
 * iOS `WorkoutLog.id`. `movement` is fully generic/gym-agnostic free
 * text — there is no fixed baseline list on either client.
 *
 * NOTE: the Phase 1 task spec described a non-optional `weight: Double`
 * field, but the current schema (as of the "generic Score field" cycle on
 * iOS) has `score`/`reps`/`sets` all optional — a member can log an
 * activity with none, some, or all of them set. This mirrors the current
 * schema instead of the stale spec.
 */
data class WorkoutLog(
    val id: String,
    val movement: String,
    val score: Double? = null,
    val reps: Int? = null,
    val sets: Int? = null,
    val dateMillis: Long
)

/**
 * The value line for this entry — whichever of [WorkoutLog.score]/[WorkoutLog.reps]/
 * [WorkoutLog.sets] were actually logged, omitting labels for fields left
 * blank. Falls back to "Logged" when none were entered (a date-only
 * session). Mirrors `WorkoutLog.formattedDetail` on iOS.
 */
val WorkoutLog.formattedDetail: String
    get() {
        return when {
            sets != null && reps != null -> {
                if (score != null) "$sets × $reps @ ${score.formatCompact()}" else "$sets × $reps"
            }
            score != null -> score.formatCompact()
            reps != null -> "$reps reps"
            sets != null -> "$sets sets"
            else -> "Logged"
        }
    }

/** Drops a trailing ".0" for whole numbers (100.0 -> "100"), same as Swift's `Double.formatted()`. */
private fun Double.formatCompact(): String =
    if (this == this.toLong().toDouble()) this.toLong().toString() else this.toString()

File: ios/Nexo/Models/WorkoutLog.swift
//
//  WorkoutLog.swift
//  Nexo
//

import Foundation

/// A single logged entry for any activity or movement — e.g. "Squat, 100kg
/// x 5 x 3 sets" or "Yoga Flow" with no score at all. Stored per-member at
/// `gyms/{gymId}/members/{userId}/workoutLogs/{id}` — not shared across the
/// gym like `GymClass`/`Member`, since this is personal training data.
/// `score`/`reps`/`sets` are all optional since not every gym type or
/// activity is meaningfully expressed as weight-lifted-for-reps.
struct WorkoutLog: Identifiable, Codable, Hashable {
    let id: String
    let movement: String
    let score: Double?
    let reps: Int?
    let sets: Int?
    let date: Date

    /// The value line for this entry:
    /// - If sets and reps are present with score: "<sets> × <reps> @ <score>"
    /// - If sets and reps are present without score: "<sets> × <reps>"
    /// - If sets or reps are not present: just the score (or reps/sets if scoreless, falling back to "Logged")
    var formattedDetail: String {
        if let sets, let reps {
            if let score {
                return "\(sets) × \(reps) @ \(score.formatted())"
            } else {
                return "\(sets) × \(reps)"
            }
        } else if let score {
            return score.formatted()
        } else if let reps {
            return "\(reps) reps"
        } else if let sets {
            return "\(sets) sets"
        } else {
            return "Logged"
        }
    }
}
